torch/testing/_internal/composite_compliance.py

```diff
@@ -0,0 +1,188 @@
+import torch
+from torch import Tensor
+import contextlib
+from typing import Iterator
+from torch.utils._pytree import tree_map
+from functools import partial
+from torch.utils._python_dispatch import enable_python_mode
+
+# TODO: move this into library proper
+@contextlib.contextmanager
+def no_dispatch() -> Iterator[None]:
+    guard = torch._C._DisableTorchDispatch()  # type: ignore[attr-defined]
+    try:
+        yield
+    finally:
+        del guard
+
+def check_attr_consistency(wrapper_tensor, metadata_name, metadata_accessor):
+    elem = wrapper_tensor.elem
+    metadata_wrapper_tensor = metadata_accessor(wrapper_tensor)
+    metadata_elem = metadata_accessor(elem)
+    if metadata_wrapper_tensor == metadata_elem:
+        return
+    raise RuntimeError(
+        f"This operator is not CompositeImplicitAutograd compliant: the "
+        f"{metadata_name} of the tensor was modified directly without "
+        f"going through the PyTorch dispatcher.")
+
+def check_metadata_consistency(wrapper_tensor):
+    if not isinstance(wrapper_tensor, CompositeCompliantTensor):
+        return
+    things_to_check = {
+        'shape': Tensor.size,
+        'dtype': lambda x: x.dtype,
+        'device': lambda x: x.device,
+        'numel': Tensor.numel
+    }
```

Comment:
@albanD @ezyang: It turns out that a wrapper tensor does not have the same strides or storage_offset as its element. Should it?

I can see this not being required for wrapper tensors like SparseTensor where strides/storage_offset make no sense. However this may be useful for something like functorch's PythonTensor that does tracing: if we trace a concrete tensor, then the wrapper should have the same strides/storage offset.

Reply:
It should.

Reply:
Thanks for the feedback. Change to wrapper object creation coming soon...